cmd/utils/cmd.go

```diff
@@ -185,32 +190,87 @@ func ExportChain(blockchain *core.BlockChain, fn string) error {
 		defer writer.(*gzip.Writer).Close()
 	}
 
-	if err := blockchain.Export(writer); err != nil {
+	var z *zip.Writer
+	if strings.HasSuffix(fn, ".zip") {
+		z = zip.NewWriter(fh)
+		writer, err = z.Create("blocks.dat")
+		if err != nil {
+			return err
+		}
+	}
+
+	log.Info("Writing blocks", "first", first, "last", last)
+	if err := blockchain.ExportN(writer, first, last); err != nil {
 		return err
 	}
-	log.Info("Exported blockchain", "file", fn)
 
+	if z != nil {
+		log.Info("Writing state", "blockNumber", last)
+		writer, err := z.Create("state.dat")
+		if err != nil {
+			return err
+		}
+		if err := exportState(blockchain, chainDb, writer, last); err != nil {
+			return err
+		}
+
+		log.Info("Writing receipts", "blockNumber", last)
+		writer, err = z.Create("receipts.dat")
+		if err != nil {
+			return err
+		}
+		if err := exportReceipts(blockchain, chainDb, writer, last); err != nil {
+			return err
+		}
+
+		err = z.Close()
+		if err != nil {
+			return err
+		}
+	}
```

Comment:
Tbh, this construct seems a bit convoluted for me. If I understand correctly, for zip files, you export everything, whilst for non-zip files you export only the blocks. Exporting only the blocks is 3 lines, but it's intermingles in between zip file handling which makes the whole thing messy.

Imho a cleaner solution would be to have a separate method that export everything and a separate one that only exports blocks. Having them share 3 loc isn't that bad imho.